### PR TITLE
[Hidden] Notify about completed report with download link via email

### LINKED ISSUE #10280: Make previously generated reports available for download
### What we should change and why (this is tech debt)

When a report takes a long time, the user may want to leave the desk and download the report later. It would be nice to tell the user that it's taking longer and provide a download link. To allow multiple downloads, we should delay report file deletion to later. We will need a cleanup task which deletes old reports for us.

#### Implementation

I think that we should leverage Active Storage for storing, downloading and deleting files.

- [x] Replace the temporary file in ReportJob with ActiveStorage::Blob ([example](https://stackoverflow.com/questions/67176634/rails-active-storage-without-model) `ActiveStorage::Blob.create_and_upload!`). The job will need to know the filename for download by the user.
- [ ] Implement a 10 second timeout in `Admin::ReportsController#show` and display a nice message with the link. A URL can be generated with `Rails.application.routes.url_helpers.url_for(blob)`. We need to check the expiry of the URL and may need to extend it.
- [x] Create an ActiveStorage::PurgeJob for each report blob: `ActiveStorage::PurgeJob.set(wait: 1.week).perform_later(blob)`.

We may want a controller action at some point which checks if the report job completed and the file is available, or if it expired already to display nicer messages to the user. The Active Storage controller will just show an error message when the report is not there.
 
### Context

Follow up from:

- #10258

Next steps:

- #10281
- #10282
- https://github.com/openfoodfoundation/ofn-install/issues/859

### Impact and timeline

We are waiting for #10258 to be thoroughly tested in production. Once all instances are running stable with that, we can remove the feature toggle and start work on this issue.

### LINKED ISSUE #10281: Email report link after lengthy generation
### What we should change and why (this is tech debt)

Notify user via email about a report that needed some time. The browser may have been closed or the request timed out.

#### Implementation

- [x] If a report took more than 5 seconds to generate then trigger an email. It should contain the link to the report.
- [x] We could use Spree::UserMailer but I would suggest creating a new `ReportMailer`.

### Context

Follow up from:

* #10280

Next steps:

* #10282
* https://github.com/openfoodfoundation/ofn-install/issues/859
* Display report link in the browser when ready

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Closes #10280
- Closes #10281 

Background reports can run for a while. If they take longer than 5 seconds, we send a download link to the user once it's complete.

We were also showing a download link in the browser but you couldn't tell when the report was done and when it was safe to click it. I removed the link there because we have a better alternative with the email.

#### What should we test?

- Enable background reports feature.
- Run a report which takes less than 5 seconds.
- It should display as usual.
- Run a report which takes more than 5 seconds.
- It should trigger an email with a download link.
- Check that report links still work 6 minutes later.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by mkllnk** [comment]
Thank you for the lovely feedback. I added this as example to the wiki. :blush:

File: app/controllers/admin/reports_controller.rb
Comment (dacook) [inline app/controllers/admin/reports_controller.rb:13]: I was wondering about this.
Yes, some servers have a long timeout, but that was only because of reports, which this PR will resolve. Then those timeouts could be adjusted to be more normal (eg 1-2min).

So why not just use the full server timeout period (less 2 secs), which the instance manager may configure to suit their needs? This would allow a better UX by providing an on-screen response rather than email.

But this change frees up workers from unnecessary waiting, which has a performance benefit. Maybe this value should be configurable so that instance managers can determine the balance between UX and performance.

But.. with a reflex we can solve the UX problem..

**Review by dacook — APPROVED** [review]
Fantastic, and I concur with JB!

I have a few queries, but I think it's all to go.

File: app/controllers/admin/reports_controller.rb
Comment (mkllnk) [reply app/controllers/admin/reports_controller.rb:13]: > with a reflex we can solve the UX problem..

Exactly, that was my conclusion. This change is to give quicker feedback to the user and free up workers. But I take your point. We could delay this change until we have the UX problem solved.

I'm not happy with the "full server timeout period (less 2 secs)" though because the two seconds are a guess. That may be more of a theoretical problem though.

Once we have a reflex, maybe we don't want to wait at all. We could display a "loading" screen straight away.

Okay, I think you convinced me that I should revert this change. We'll solve it with the reflex. Low number of workers has not been a reported problem and if that's because of too many reports then it may be good to limit that...

File: spec/system/admin/reports_spec.rb
Comment (mkllnk) [reply spec/system/admin/reports_spec.rb]: Hehe, I was wondering if I should have committed this separately. This line is only needed when you run the spec on its own. Then there are other `fetch` calls as well. This line allows these other calls. Otherwise Rspec thinks that our expected call should be the only one.

That line should have been in the original commit of the next line. But that was a different PR and I didn't notice that the spec didn't run on its own then.

File: spec/system/admin/reports_spec.rb
Comment (mkllnk) [reply spec/system/admin/reports_spec.rb:80]: The browser is not looking at the email. I tried to replace the HTML document with the email body and then use the browser. But while everything displayed, the browser couldn't find the link. There must be some DOM caching that wasn't updated by `document.write(html)`.

Oh and that `Capybara::Node::Simple` doesn't support actions, only inspection.

I also tried to `visit report_link` but that triggered a download. While that's cool and testable, the `visit` call took a minute to return. So in that scenario the test execution time went from 5 seconds to 2 minutes and 5 seconds and the report was still not "on screen". That may be a bug in the driver. I'm not sure.

Comment (dacook) [reply spec/system/admin/reports_spec.rb:80]: > I also tried to `visit report_link` but that triggered a download.

Yep, that's what I meant, but that's strange it takes a minute, which makes this method much better 👍

File: app/controllers/admin/reports_controller.rb
Comment (dacook) [reply app/controllers/admin/reports_controller.rb:13]: Haha sorry I didn't meant to convince you to change it now. But yes I had the same conclusion: if we can push a response on completion with a reflex, then we don't need to wait here at all.

**Comment by mkllnk** [comment]
I dropped the last commit reducing the timeout to 10 seconds. Let's solve it properly with a reflex. Moving to Test Ready.

**Comment by drummer83** [comment]
@mkllnk Just a quick feedback on this one...
After reading through the issues and the PR I was expecting that the email notification would be shown on screen after 5 seconds. However on staging UK it appeared after 2 minutes only, which is the timeout setting, I guess? Is that what you would expect?

image

And on staging AU I get an error 500 Internal Server Error after 2 minutes. 
Here are some Bugsnags which may be related:
https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/63d7576eb3baf80009f92e76?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging
https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/63e28a6f76fcc500082f6ab3?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging
https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/6449660caa52d20008c52081?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging
https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/6449557e99817e0008c84da3?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging

Would be interesting to hear your thoughts! Thanks!

## Before staging the PR

| Report type | Background reports activated | **admin** user (**w/o** date restriction) | **admin** user (**with** date restriction) | **super admin** user (**w/o** date restriction) | **super admin** user (**with** date restriction) | 
| - | - | - | - | - | - |
| Orders And Distributors | yes | :heavy_check_mark: | not tested | Download link | not tested |
| Bulk Co-op Supplier Report | yes | :heavy_check_mark: | not tested | Download link | not tested |
| Payments By Type | yes | :heavy_check_mark: | not tested | Download link | not tested |
| Order Cycle Supplier Totals | yes | :heavy_check_mark: | not tested | :x: Internal Server Error 500 (19:11) | :heavy_check_mark: |
| Mailing List | yes | :heavy_check_mark: | not tested | :heavy_check_mark: | n.a. |
| All products | yes | :heavy_check_mark: | not tested | :x: Timeout 504 | n.a. |
| Users & Enterprises | yes | :heavy_check_mark: | not tested | :x: Internal Server Error 500 (19:20) | n.a. |
| Enterprise Fee Summary | yes | :heavy_check_mark: | not tested | :heavy_check_mark: | not tested |
| Payment Methods Report | yes | :heavy_check_mark: | not tested | :x: Timeout 504 | :heavy_check_mark: |
| Tax Types | yes | :heavy_check_mark: | not tested | :x: Timeout 504 | not tested |
|  Xero Summary | yes | :heavy_check_mark: | not tested | :x: Timeout 504 | not tested |
| Pack By Customer | yes | :heavy_check_mark: | not tested | :x: Internal Server Error 500 | not tested |
| Revenues By Hub | yes | n.a. | n.a. | :heavy_check_mark: | :heavy_check_mark:  |

## Notes to myself
- close browser session
- click link after more than 6 minutes

**Comment by filipefurtad0** [comment]
From the bugsnag errors, several reports trigger the error in staging-AU:

`incompatible character encodings: UTF-8 and ASCII-8BIT`
```

/reports/bulk_coop
/reports/packing
/reports/users_and_enterprises
/reports/orders_and_fulfillment
```

However, from your notes I get the idea that this error was not triggered in staging-UK at all. This could suggest it is related with the data itself.

**Comment by sigmundpetersen** [comment]
> incompatible character encodings: UTF-8 and ASCII-8BIT

Also reported in #10758

**Review by Matt-Yorkley — COMMENTED** [review]
Have you tried this with a report using a huge amount of production data?

Timeout values on some servers range from `120` to `360`, right? In that case `#max_wait_time` will return `118` or `358`, and the controller will be stuck in a `#sleep` loop here for that many seconds before responding? Yikes! :grimacing: 

```ruby
Timeout.timeout(max_wait_time) do
  sleep 1 until @blob.content_stored?
end
```
That will mean a puma thread is held up for that length of time while serving the request, which kind of defeats the purpose of passing the workload off to a background job, right? I would not recommend putting this code into production as-is...

On a more positive note, there's a much nicer way of solving this problem and avoiding calling `sleep` by using `cable_ready`, but we don't have any examples of it in the codebase yet. Essentially you can push a live DOM update to the user from inside of the `ReportJob` itself, and the controller doesn't have to wait at all here before serving it's response. I could whip you up an example if you're interested?

As a completely separate side-note; I think the performance issues with reports could probably be fixed, which would mean that passing them off to a background job might not be necessary in the first place, but that's a different story...

**Comment by mkllnk** [comment]
Hi Matt, nice to see you here.

> controller will be stuck in a #sleep loop here 

Before it was busy compiling a report. Why is sleeping for several minutes worse than being busy for several minutes?

> there's a much nicer way of solving this problem and avoiding calling sleep by using cable_ready

Yes, exactly. That's our plan for the next step. This was just one iteration which I hope is better than it was before. Here's the cable_ready issue:

* #10752

> I think the performance issues with reports could probably be fixed

I think so, too. You started excellent work on moving the grouping into SQL. We should make more use of that. But the question is if there will always be some reports that are taking longer. If we want to reduce our Puma timeout to a sane value then we need to run the reports in the background anyway.

Initially we thought that background reports would be a "quick fix" but it's turning out to be more work than we thought.

**Comment by mkllnk** [comment]
@drummer83 

> After reading through the issues and the PR I was expecting that the email notification would be shown on screen after 5 seconds. 

I'm sorry, that was the case in an older version of the PR. I removed that change though because we thought it's better to display the report when possible instead of just a message and then needing to look at your emails.

**Comment by drummer83** [comment]
@mkllnk @Matt-Yorkley 
How shall we proceed? Should we move back to In Dev? Or shall I finish testing?

**Comment by Matt-Yorkley** [comment]
I just had a play with it and I can see what you're aiming for. A couple of thoughts:

- Every time the form is submitted it creates a new file. That could add up pretty quick?
- There's no way to view/browse/delete previously generated reports. It needs a decent new UI for showing them.
- The email part isn't opt-in or opt-out, it just happens?
- Adding in-app user notifications is pretty easy. It might be useful for some of these UX conundrums.
- The flow from submitting the form to seeing the results needs some thought. Like: submit the form -> run the job -> hide the filters? -> show a loading indicator? -> update the page when the job completes?
- If the format is HTML the email link doesn't really work, it shows an unstyled table on a blank page. We need to be able to send a link that will load the blob's content and show it nicely on the screen, and think how that fits in to the index/show/delete type actions for any/all previously generated reports.

**Comment by mkllnk** [comment]
@drummer83 

> How shall we proceed? Should we move back to In Dev? Or shall I finish testing?

I moved it back to In Dev until we the fix for the report crashes is merged:

* #10762

**Comment by mkllnk** [comment]
@Matt-Yorkley 

> Every time the form is submitted it creates a new file. That could add up pretty quick?

Yes, it could. And we keep them for one month. Storage on S3 is pretty cheap. But we need to monitor instances with local disk storage.

We could rethink that retention and delete the files straight away unless an email with a link has been sent out.

With cable_ready, we may have some more options. It would notify about the job completion which means that we don't need to synchronise over the Active Storage blob. Then we can only store if needed and render straight away otherwise.

> There's no way to view/browse/delete previously generated reports. It needs a decent new UI for showing them.

Correct, I was imagining that as well. That's more work though. First priority was to solve the memory issue, not improving UX.

> The email part isn't opt-in or opt-out, it just happens?

Yes, at this stage. We could include a checkbox on the page to make it an option.

> Adding in-app user notifications is pretty easy. It might be useful for some of these UX conundrums.

Yes, that's planned.

> The flow from submitting the form to seeing the results needs some thought.

Agreed.

> If the format is HTML the email link doesn't really work, it shows an unstyled table on a blank page.

It's better than a server error. But yes, a whole UI around generated reports would be nice.

**Comment by Matt-Yorkley** [comment]
Cool :+1:

Just for future reference; if you're ever thinking of making any kind of feature that involves notifications of any kind, the **Noticed** gem is absolutely the way to go: https://github.com/excid3/noticed. `Noticed` is to notifications what `ActionJob` is to jobs, or what `ActionMailer` is to emails. It's incredibly good.

**Comment by mkllnk** [comment]
Nice one, thanks.

**Comment by Matt-Yorkley** [comment]
So... I made you nice PR for this @mkllnk but it's partly blocked by `jQuery` and `select2` interfering with (and breaking) the other javascript I'm trying to use. We need to get rid of `select2` from all of the select elements on all the filters on every report... :grimacing:

**Comment by Matt-Yorkley** [comment]
Alright, I managed to work around some of the legacy javascript issues :tada: 

I made a little PR for shifting the feedback into `CableReady`, pull it and give it a try: https://github.com/mkllnk/openfoodnetwork/pull/1

**Comment by mkllnk** [comment]
I just tested this on au-staging and found another issue. Apparently we are limited to 1 week when setting the link expiry.

```
2023-05-05T04:10:45.426Z pid=16934 tid=3wk2 class=ActionMailer::MailDeliveryJob jid=b8cc286df1d7c2852456de7a ERROR: Error performing ActionMailer::Mai
lDeliveryJob (Job ID: ba1a053b-55bc-4a8f-84dd-9fb322df519a) from Sidekiq(default) in 8.16ms: ActionView::Template::Error (expires_in value of 2629746 
exceeds one-week maximum.):
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/aws-sdk-s3-1.121.0/lib/aws-sdk-s3/presigner.rb:157:in `expires_in'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/aws-sdk-s3-1.121.0/lib/aws-sdk-s3/presigner.rb:137:in `_presigned_request'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/aws-sdk-s3-1.121.0/lib/aws-sdk-s3/presigner.rb:78:in `presigned_url'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/aws-sdk-s3-1.121.0/lib/aws-sdk-s3/customizations/object.rb:215:in `presigned_url'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activestorage-7.0.4.3/lib/active_storage/service/s3_service.rb:118:in `private_url'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activestorage-7.0.4.3/lib/active_storage/service.rb:123:in `block in url'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activesupport-7.0.4.3/lib/active_support/notifications.rb:206:in `block in instrument'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activesupport-7.0.4.3/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activesupport-7.0.4.3/lib/active_support/notifications.rb:206:in `instrument'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activestorage-7.0.4.3/lib/active_storage/service.rb:163:in `instrument'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activestorage-7.0.4.3/lib/active_storage/service.rb:118:in `url'
/home/openfoodnetwork/.gem/ruby/3.0.0/gems/activestorage-7.0.4.3/app/models/active_storage/blob.rb:217:in `url'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-05-05-035232/app/models/report_blob.rb:38:in `expiring_service_url'
```

**Comment by mkllnk** [comment]
This is ready for testing again.

**Comment by drummer83** [comment]
Hi @mkllnk,
Thanks for this round of updates and the additional information!
I am afraid we still have some issues here. Please have a look at my 'observations' chapter below.

## After staging the PR

| Report type | Background reports activated | **admin** user (**w/o** date restriction) | **admin** user (**with** date restriction) | **super admin** user (**w/o** date restriction) | **super admin** user (**with** date restriction) | 
| - | - | - | - | - | - |
| Orders And Distributors | yes | on screen (15 s) :heavy_check_mark: | not tested | AU: email:question: UK: email:question: | not tested |
| Bulk Co-op Supplier Report | yes | not tested | not tested | AU: email:question: UK: email:heavy_check_mark: | not tested |
| Payments By Type | yes | not tested | not tested | AU: email:question: UK: email:question: | not tested |
| Order Cycle Supplier Totals | yes | on screen | not tested | not tested | not tested |
| Mailing List | yes | not tested | not tested | AU: on screen:heavy_check_mark: + email:x: UK: email:question: | n.a. |
| All products | yes | not tested | not tested | AU: email:heavy_check_mark: UK: email:question: | n.a. |
| Users & Enterprises | yes | not tested | not tested | AU: email:heavy_check_mark: UK: email:question: | n.a. |
| Enterprise Fee Summary | yes | not tested | not tested | not tested | not tested |
| Payment Methods Report | yes | not tested | not tested | not tested | not tested |
| Tax Types | yes | not tested | not tested | not tested | not tested |
|  Xero Summary | yes | not tested | not tested | not tested | not tested |
| Pack By Customer | yes | not tested | not tested | not tested | not tested |
| Revenues By Hub | yes | n.a. | n.a. | not tested | not tested  |

## Notes to myself
- close browser session
- click link after more than 6 minutes

## Observations
- In staging AU every report generated before timeout sends an email as well (tested on screen and pdf). :x:
- I am receiving no emails from staging UK (added myself as 'send a copy of all mails to' - same as AU) at all. The test mail is sent and received correctly. Access to the staging email account (hello@openfoodnetwork.org.uk) could help. :question: 
- Any user has access to the downloaded file (if the link is known). Has this been considered to be safe enough because it's very hard to guess? :question: 
- If I select pdf as output and the generation takes longer than the timeout (2 min) the email is sent correctly, but on screen there is an error 504 Gateway Timeout. Probably the same for csv and xlsx? :x:
- The text in the email notification reads: The link below will expire after one month. I believe this was changed to one week, due to S3 limitations? Even though this could be changed in Transifex (probably) I think we should get it right from the beginning - assuming that we need to implement changes due to the other findings above anyway. :x:

I believe we need to fix some things before finalizing the testing and merging. I will move this back to In Dev for now.
Thanks!

**Comment by mkllnk** [comment]
Thank you for the thorough test again.

> In staging AU every report generated before timeout sends an email as well (tested on screen and pdf). x

The email is supposed to be sent after 5 seconds. Did you mean that with timeout or did you mean the longer request timeout?

>     I am receiving no emails from staging UK (added myself as 'send a copy of all mails to' - same as AU) at all. The test mail is sent and received correctly. Access to the staging email account ([hello@openfoodnetwork.org.uk](mailto:hello@openfoodnetwork.org.uk)) could help. question

That's weird. The logs show that you ran 5 background reports. Two of them took longer than 5 seconds. One 5.7 seconds and the other one 512 seconds = 8.5 minutes. They should have triggered emails.

The test email is not a good indicator because it sends emails directly, not via background jobs. What about order confirmations?

Hm, your testing notes do say that you got one email for the Bulk Co-op Supplier Report though. I'm not sure what to make of that.

>     Any user has access to the downloaded file (if the link is known). Has this been considered to be safe enough because it's very hard to guess? question

Yes, the link gives access to the file. It's nearly impossible to guess such a link.

>     If I select pdf as output and the generation takes longer than the timeout (2 min) the email is sent correctly, but on screen there is an error 504 Gateway Timeout. Probably the same for csv and xlsx? x

Interesting. That means that the setup of the report parameters took longer than two seconds. I wouldn't worry about this too much because Matt has a PR to replace this timeout mechanism with a much better alternative. We can ignore this.

>     The text in the email notification reads: The link below will expire after one month. I believe this was changed to one week, due to S3 limitations? Even though this could be changed in Transifex (probably) I think we should get it right from the beginning - assuming that we need to implement changes due to the other findings above anyway. x

You are totally right. I'll change it now and move it back to test-ready. I hope that I clarified the questions and that the emails are working correctly.

**Comment by drummer83** [comment]
Hi @mkllnk,
I'm so sorry that this PR is a chewing gum (Can you say that in English?) and it is getting frustrating for you. Still I have difficulties to judge if this is working as expected or not.

> > In staging AU every report generated before timeout sends an email as well (tested on screen and pdf). x
> 
> The email is supposed to be sent after 5 seconds. Did you mean that with timeout or did you mean the longer request timeout?

Now I am totally confused... How does this correlate to your answer [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/10732#issuecomment-1526838186)? What should happen after 5 seconds? What after 2 minutes?

This is what I would expect (maybe):

| Report generation duration | On screen | Send email notification |
| - | - | - |
| t < 5 seconds | display report | no |
| 5 seconds < t < 2 minutes | after 5 s: display email hint :question: | not yet |
| | when report ready: display report :question: | yes :question: |
| t > 2 minutes | after 2 min: still email hint | not yet |
| | when report ready: still email hint? | yes |

Can you please fill in the table with what you/the product team would expect?

> > I am receiving no emails from staging UK (added myself as 'send a copy of all mails to' - same as AU) at all. The test mail is sent and received correctly. Access to the staging email account ([hello@openfoodnetwork.org.uk](mailto:hello@openfoodnetwork.org.uk)) could help. question
> 
> That's weird. The logs show that you ran 5 background reports. Two of them took longer than 5 seconds. One 5.7 seconds and the other one 512 seconds = 8.5 minutes. They should have triggered emails.
> 
> The test email is not a good indicator because it sends emails directly, not via background jobs. What about order confirmations?
> 
> Hm, your testing notes do say that you got one email for the Bulk Co-op Supplier Report though. I'm not sure what to make of that.

Oh, you're right! I did receive one email. I will check again.

> > The text in the email notification reads: The link below will expire after one month. I believe this was changed to one week, due to S3 limitations? Even though this could be changed in Transifex (probably) I think we should get it right from the beginning - assuming that we need to implement changes due to the other findings above anyway. x
> 
> You are totally right. I'll change it now and move it back to test-ready. I hope that I clarified the questions and that the emails are working correctly.

Thanks! :pray:

**Comment by Matt-Yorkley** [comment]
Not sure if this helps resolve the impasse or makes it more complicated (hopefully the former), but... I think that sending the email after 5 seconds isn't actually needed at all if we're giving the user feedback via CableReady as soon as the job finishes :point_right: https://github.com/mkllnk/openfoodnetwork/pull/1.

**Comment by mkllnk** [comment]
Thank you for the table @drummer83. I'll use it to explain what I expect:


| Report generation duration | On screen | Send email notification |
| - | - | - |
| t < 5 seconds | display report | no |
| 5 seconds < t < 2 minutes | when report ready: display report | yes |
| t > 2 minutes | after 2 min: display email hint (this may fail on slow servers, ignore) | not yet |
| | when report ready: still email hint as above | yes |


So the two changes in this PR are:

* Email the report when ready if it took longer than 5 seconds.
* Changed text in the timeout message: mention email, don't show broken link

**Comment by mkllnk** [comment]
> I think that sending the email after 5 seconds isn't actually needed at all if we're giving the user feedback via CableReady as soon as the job finishes

Yes, it's not necessary. I still think that it's a useful feature though. It would be great if we had a checkbox for it though so that the user can decide to get an email or not.

**Comment by drummer83** [comment]
Hi @mkllnk, 
Nope, sorry, still no success. Maybe I should give up and ask @filipefurtad0 to take over or get his advice.

### Staging AU
This time Staging AU is not sending emails at all. No order confirmations, no password reset emails, no report ready emails. :cry:  Test email from super-admin configuration is failing as well. Can you fix this and maybe let me know what I could have done?

### Staging UK
On staging UK emails are working and I could confirm most parts of the expected behavior:

| Report generation duration | On screen | Send email notification |
| - | - | - |
| t < 5 seconds | display report :heavy_check_mark:  | no :heavy_check_mark: |
| 5 seconds < t < 2 minutes | when report ready: display report :heavy_check_mark: | yes :heavy_check_mark: |
| t > 2 minutes | after 2 min: display email hint (this may fail on slow servers, ignore) :heavy_check_mark: | not yet :heavy_check_mark: |
| | when report ready: still email hint as above :x: | yes :x: |

I never received the email for times > 2 minutes. Maybe it's because the report never finished but I have waited for one hour. Could it be that the report generation is aborted after a certain timeout? 

### Staging FR
On staging FR it's the same as on staging UK and I increased the date range step by step to make sure that I was not trying to generate an hours-long report at the first try.

### UX
I will create a wishlist item to collect some ideas for UX improvements around this feature. Now that I have been testing it, it's still fresh on my mind.

**Comment by mkllnk** [comment]
Okay, so in general notification emails are sent but you didn't receive one after two minutes. Looking at Bugsnag, there seems to be an unrelated error with report data or a bug in the reports: String can't be coerced into BigDecimal
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/643fc2072915e90008ca984e

That's in OrderCycleSupplierTotals. Did you test that one?
Since it happened beyond the timeout, you wouldn't have been able to test it with normal foreground reports.

I think I still see this as an improvement and I will merge this PR and look at the issue with the report data separately. It also paves the way for Matt's PR which should give you a much better UX.

**Comment by drummer83** [comment]
> That's in OrderCycleSupplierTotals. Did you test that one?

Yes, I did. :+1: 

Sorry, not my best testing performance. I should have checked Bugsnag and tried another report type.

**Comment by mkllnk** [comment]
I found that the bug was in that one report and only in the summary row. :see_no_evil: That was really bad luck on your part. But I'm glad that you found another bug! :+1:

**Pull request merged**
